extensions/amp-story-auto-ads/0.1/story-ad-page.js

```diff
@@ -0,0 +1,486 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  AnalyticsEvents,
+  AnalyticsVars,
+  STORY_AD_ANALYTICS,
+} from './story-ad-analytics';
+import {CommonSignals} from '../../../src/common-signals';
+import {CtaTypes} from './story-ad-localization';
+import {assertConfig} from '../../amp-ad-exit/0.1/config';
+import {assertHttpsUrl} from '../../../src/url';
+import {CSS as attributionCSS} from '../../../build/amp-story-auto-ads-attribution-0.1.css';
+import {
+  createElementWithAttributes,
+  elementByTag,
+  isJsonScriptTag,
+  iterateCursor,
+  openWindowDialog,
+  toggleAttribute,
+} from '../../../src/dom';
+import {createShadowRootWithStyle} from '../../amp-story/1.0/utils';
+import {dev, user, userAssert} from '../../../src/log';
+import {dict} from '../../../src/utils/object';
+import {getA4AMetaTags, getFrameDoc} from './utils';
+import {getServicePromiseForDoc} from '../../../src/service';
+import {parseJson} from '../../../src/json';
+import {setStyles} from '../../../src/style';
+
+/** @const {string} */
+const TAG = 'amp-story-auto-ads:page';
+
+/** @const {number} */
+const TIMEOUT_LIMIT = 10000; // 10 seconds
+
+/** @const {string} */
+const GLASS_PANE_CLASS = 'i-amphtml-glass-pane';
+
+/** @enum {string} */
+const PageAttributes = {
+  LOADING: 'i-amphtml-loading',
+  IFRAME_BODY_VISIBLE: 'amp-story-visible',
+};
+
+/** @enum {string} */
+const DataAttrs = {
+  CTA_TYPE: 'data-vars-ctatype',
+  CTA_URL: 'data-vars-ctaurl',
+};
+
+/** @enum {string} */
+const A4AVarNames = {
+  ATTRIBUTION_ICON: 'attribution-icon',
+  ATTRIBUTION_URL: 'attribution-url',
+  CTA_TYPE: 'cta-type',
+  CTA_URL: 'cta-url',
+};
+
+export class StoryAdPage {
+  /**
+   * @param {!Element} storyAutoAdsEl
+   * @param {!JsonObject} config
+   * @param {number} index
+   * @param {!./story-ad-localization.StoryAdLocalization} localization
+   */
+  constructor(storyAutoAdsEl, config, index, localization) {
+    /** @private @const {!JsonObject} */
+    this.config_ = config;
+
+    /** @private @const {number} */
+    this.index_ = index;
+
+    /** @private @const {!./story-ad-localization.StoryAdLocalization} */
+    this.localizationService_ = localization;
+
+    /** @private @const {string} */
+    this.id_ = `i-amphtml-ad-page-${this.index_}`;
+
+    /** @private @const {!Window} */
+    this.win_ = storyAutoAdsEl.getAmpDoc().win;
+
+    /** @private @const {!Document} */
+    this.doc_ = this.win_.document;
+
+    /** @private @const {!Promise} */
+    this.analytics_ = getServicePromiseForDoc(
+      storyAutoAdsEl,
+      STORY_AD_ANALYTICS
+    );
+
+    /** @private {?number} */
+    this.timeCreated_ = null;
+
+    /** @private {?Element} */
+    this.pageElement_ = null;
+
+    /** @private {?Element} */
+    this.adElement_ = null;
+
+    /** @private {?Document} */
+    this.adDoc_ = null;
+
+    /** @private {?string} */
+    this.ctaText_ = null;
```

Comment:
let's try to reduce the number of member fields. for example, this one can be a local var.

Reply:
moved to local var.